File: apps/ios/Shared/Views/Badges/BadgesRedeemCodeView.swift
//
//  BadgesRedeemCodeView.swift
//  SimpleX (iOS)
//
//  Created by spaced4ndy on 28.07.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat
import CodeScanner

private let badgeCodePrefix = "SB"
private let badgeCodeBodyLength = 20
private let badgeCodeGroupLength = 5

// Regroups what was typed; validity and the folding of ambiguous characters are core's alone.
private func formatBadgeCodeInput(_ s: String) -> String {
    var normalized = ""
    for c in s.prefix(256).uppercased() {
        guard c.isLetter || c.isNumber else { continue }
        normalized.append(c)
        if normalized.count == badgeCodePrefix.count + badgeCodeBodyLength { break }
    }
    guard normalized.hasPrefix(badgeCodePrefix) else { return normalized }
    var groups = [badgeCodePrefix]
    var i = normalized.index(normalized.startIndex, offsetBy: badgeCodePrefix.count)
    while i < normalized.endIndex {
        let j = normalized.index(i, offsetBy: badgeCodeGroupLength, limitedBy: normalized.endIndex) ?? normalized.endIndex
        groups.append(String(normalized[i..<j]))
        i = j
    }
    return groups.joined(separator: "-")
}

enum BadgeRedeemOutcome {
    case redeemed
    case refused(message: String)
    case cancelled
}

// sets the badge before returning, so a caller that dismisses on .redeemed lands on Your Badge
// instead of showing the switch from Support. .cancelled when the user cancels the retry alert.
func redeemBadgeCode(_ user: User, _ code: String) async -> BadgeRedeemOutcome {
    do {
        guard let redeemed = try await apiRedeemBadgeCode(user.userId, code) else { return .cancelled }
        return await MainActor.run { () -> BadgeRedeemOutcome in
            BadgeModel.shared.set(userId: user.userId, badgeState: redeemed.badgeState)
            ChatModel.shared.updateUser(redeemed.user)
            if let badgeState = redeemed.badgeState, !badgeState.shown {
                // a replay adds no purchase; a fresh code's badge can be retired on arrival
                return .refused(message: redeemed.newBadge
                    ? NSLocalizedString("The code was accepted, but the badge it grants has already ended.", comment: "alert message")
                    : NSLocalizedString("This code has already been used.", comment: "alert message")
                )
            }
            UserDefaults.standard.set(true, forKey: DEFAULT_SUPPORTER_BANNER_SHOWN)
            return .redeemed
        }
    } catch let error {
        logger.error("apiRedeemBadgeCode: \(responseError(error))")
        return .refused(message: redeemErrorText(error))
    }
}

func showCannotRedeemAlert(_ message: String) {
    showAlert(NSLocalizedString("Cannot redeem code", comment: "alert title"), message: message)
}

struct BadgesRedeemCodeView: View {
    @EnvironmentObject var theme: AppTheme
    @EnvironmentObject var chatModel: ChatModel
    @Environment(\.dismiss) var dismiss: DismissAction
    @State private var code = ""
    @State private var canonicalCode: String? = nil
    @State private var submitting = false
    @State private var showQRCodeScanner = true

    var body: some View {
        ZStack {
            GeometryReader { g in
                ScrollView {
                    VStack(alignment: .center, spacing: 16) {
                        entryContent(g)
                    }
                    .padding(.horizontal, 25)
                    .padding(.top, 8)
                    .padding(.bottom, 20)
                    .frame(minHeight: g.size.height)
                }
            }

            if submitting {
                ZStack {
                    Circle()
                        .fill(.white)
                        .opacity(0.7)
                        .frame(width: 56, height: 56)
                    ProgressView().scaleEffect(2)
                }
            }
        }
        .frame(maxHeight: .infinity)
        .navigationBarTitleDisplayMode(.inline)
    }

    @ViewBuilder
    private func entryContent(_ g: GeometryProxy) -> some View {
        Text("Redeem code")
            .font(.largeTitle)
            .bold()
            .foregroundColor(theme.colors.primary)
            .multilineTextAlignment(.center)
            .fixedSize(horizontal: false, vertical: true)

        Text("Paste the code you received.")
            .font(.body)
            .multilineTextAlignment(.center)
            .fixedSize(horizontal: false, vertical: true)

        codeField()

        pasteButton()

        ScannerInView(
            showQRCodeScanner: $showQRCodeScanner,
            scannerPaused: $submitting,
            processQRCode: processQRCode,
            scanMode: .oncePerCode,
            placeholderBackground: Color(.tertiarySystemFill)
        )
        .fixedSize(horizontal: false, vertical: true)
        .padding(.top, 12)

        Spacer(minLength: 0)

        VStack(spacing: 10) {
            submitButton()
                .padding(.vertical, 10)
            Color.clear
                .frame(height: 22)
        }
        .padding(.bottom, g.safeAreaInsets.bottom == 0 ? 20 : 0)
    }

    private func codeField() -> some View {
        TextField(text: $code) { Text(verbatim: "SB-XXXXX-XXXXX-XXXXX-XXXXX") }
            .font(.body.monospaced())
            .multilineTextAlignment(.center)
            .autocorrectionDisabled(true)
            .textInputAutocapitalization(.characters)
            .disabled(submitting)
            .padding(EdgeInsets(top: 14, leading: 12, bottom: 14, trailing: 12))
            .background(Color(.tertiarySystemFill))
            .cornerRadius(10.0)
            .onChange(of: code) { applyCodeInput($0) }
    }

    private func pasteButton() -> some View {
        Button {
            if let pasted = UIPasteboard.general.string { applyCodeInput(pasted) }
        } label: {
            Text("Paste")
                .font(.body)
                .fontWeight(.medium)
                .foregroundColor(theme.colors.primary)
        }
        .disabled(submitting)
    }

    // a changed field re-enters through onChange with the formatted text, which is when it is parsed
    private func applyCodeInput(_ s: String) {
        let formatted = formatBadgeCodeInput(s)
        if formatted != code {
            code = formatted
        } else {
            canonicalCode = parseBadgeCode(formatted)
        }
    }

    private func processQRCode(_ resp: Result<ScanResult, ScanError>) {
        switch resp {
        case let .success(r):
            let formatted = formatBadgeCodeInput(r.string)
            if let canonical = parseBadgeCode(formatted) {
                // set here rather than through onChange, which runs after redeem() reads it
                code = formatted
                canonicalCode = canonical
                redeem()
            } else {
                showAlert(
                    NSLocalizedString("Invalid QR code", comment: "alert title"),
                    message: NSLocalizedString("The code you scanned is not a badge code.", comment: "alert message")
                )
            }
        case let .failure(e):
            logger.error("processQRCode QR code error: \(e.localizedDescription)")
            showAlert(
                NSLocalizedString("Invalid QR code", comment: "alert title"),
                message: String.localizedStringWithFormat(NSLocalizedString("Error scanning code: %@", comment: "alert message"), e.localizedDescription)
            )
        }
    }

    private func submitButton() -> some View {
        let disabled = canonicalCode == nil || submitting
        return Button {
            redeem()
        } label: {
            Text("Redeem")
        }
        .buttonStyle(OnboardingButtonStyle(isDisabled: disabled))
        .disabled(disabled)
    }

    private func redeem() {
        guard let sending = canonicalCode, let user = chatModel.currentUser else { return }
        submitting = true
        Task {
            let outcome = await redeemBadgeCode(user, sending)
            await MainActor.run {
                submitting = false
                switch outcome {
                case .redeemed: dismiss()
                case let .refused(message): showCannotRedeemAlert(message)
                case .cancelled: break
                }
            }
        }
    }
}

func openBadgeLink(_ codeText: String) {
    guard let code = parseBadgeCode(codeText) else {
        return showCannotRedeemAlert(NSLocalizedString("This code is not valid.", comment: "alert message"))
    }
    showAppSheet {
        NavigationView {
            BadgesRedeemLinkView(code: code)
                .modifier(ThemedBackground())
        }
    }
}

enum BadgeLinkStep {
    case confirming
    case issuing
    case redeemed
}

// Any web page or chat message can send a badge link, and a profile holds one badge at a time,
// so this screen asks before redeeming, names the profile, and offers nothing but the redemption.
struct BadgesRedeemLinkView: View {
    @EnvironmentObject var theme: AppTheme
    @EnvironmentObject var chatModel: ChatModel
    let code: String
    @State private var step = BadgeLinkStep.confirming

    var body: some View {
        switch step {
        case .confirming: confirming()
        case .issuing: beingIssued()
        case .redeemed: BadgesView(showsAsSheet: true)
        }
    }

    private func confirming() -> some View {
        VStack(alignment: .center, spacing: 16) {
            Text("Add badge to your profile?")
                .font(.largeTitle)
                .bold()
                .foregroundColor(theme.colors.primary)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            Text(String.localizedStringWithFormat(NSLocalizedString("The badge will be added to the profile %@.", comment: "badge link confirmation"), chatModel.currentUser?.displayName ?? ""))
                .font(.body)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            Spacer()

            VStack(spacing: 10) {
                Button {
                    redeemFromLink()
                } label: {
                    Text("Add badge")
                }
                .buttonStyle(OnboardingButtonStyle(isDisabled: false))
                .padding(.vertical, 10)

                Button {
                    dismissAllSheets()
                } label: {
                    Text("Cancel")
                        .font(.body)
                        .fontWeight(.medium)
                        .foregroundColor(theme.colors.primary)
                }
                .frame(height: 22)
            }
        }
        .padding(.horizontal, 25)
        .padding(.top, 48)
        .padding(.bottom, 20)
        .frame(maxHeight: .infinity)
        .navigationBarTitleDisplayMode(.inline)
    }

    private func beingIssued() -> some View {
        VStack(alignment: .center, spacing: 16) {
            Text("Badge is being issued")
                .font(.largeTitle)
                .bold()
                .foregroundColor(theme.colors.primary)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            Spacer()

            ProgressView().scaleEffect(2)

            Spacer()
        }
        .padding(.horizontal, 25)
        .padding(.top, 48)
        .padding(.bottom, 20)
        .frame(maxHeight: .infinity)
        .navigationBarTitleDisplayMode(.inline)
    }

    private func redeemFromLink() {
        guard let user = chatModel.currentUser else { return dismissAllSheets() }
        step = .issuing
        Task {
            let outcome = await redeemBadgeCode(user, code)
            await MainActor.run {
                switch outcome {
                case .redeemed: step = .redeemed
                case let .refused(message): dismissAllSheets { showCannotRedeemAlert(message) }
                case .cancelled: dismissAllSheets()
                }
            }
        }
    }
}

struct BadgesRedeemCodeView_Previews: PreviewProvider {
    static var previews: some View {
        NavigationView {
            BadgesRedeemCodeView()
        }
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/badges/BadgesRedeemCodeView.kt
package chat.simplex.common.views.badges

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import dev.icerock.moko.resources.compose.stringResource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import chat.simplex.common.model.*
import chat.simplex.common.model.ChatController.appPrefs
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.helpers.*
import chat.simplex.common.views.newchat.QRCodeScanner
import chat.simplex.common.views.onboarding.OnboardingActionButton
import chat.simplex.common.views.onboarding.TextButtonBelowOnboardingButton
import chat.simplex.res.MR

private const val badgeCodePrefix = "SB"
private const val badgeCodeBodyLength = 20
private const val badgeCodeGroupLength = 5

// Regroups what was typed; validity and the folding of ambiguous characters are core's alone.
private fun formatBadgeCodeInput(s: String): String {
  val normalized = StringBuilder()
  for (c in s.take(256).uppercase()) {
    if (!c.isLetterOrDigit()) continue
    normalized.append(c)
    if (normalized.length == badgeCodePrefix.length + badgeCodeBodyLength) break
  }
  if (!normalized.startsWith(badgeCodePrefix)) return normalized.toString()
  val groups = mutableListOf(badgeCodePrefix)
  var i = badgeCodePrefix.length
  while (i < normalized.length) {
    val j = minOf(i + badgeCodeGroupLength, normalized.length)
    groups.add(normalized.substring(i, j))
    i = j
  }
  return groups.joinToString("-")
}

sealed class BadgeRedeemOutcome {
  object Redeemed: BadgeRedeemOutcome()
  class Refused(val message: String): BadgeRedeemOutcome()
  object Cancelled: BadgeRedeemOutcome()
}

// sets the badge before returning, so a caller that dismisses on Redeemed lands on Your Badge
// instead of showing the switch from Support. Cancelled when the user cancels the retry alert.
suspend fun redeemBadgeCode(rhId: Long?, user: User, code: String): BadgeRedeemOutcome =
  when (val result = chatModel.controller.apiRedeemBadgeCode(rhId, user.userId, code)) {
    null -> BadgeRedeemOutcome.Cancelled
    is BadgeRedeemResult.Redeemed -> withContext(Dispatchers.Main) {
      val badgeState = result.badgeState
      BadgeModel.set(rhId, user.userId, badgeState)
      chatModel.updateUser(result.user)
      if (badgeState != null && !badgeState.shown) {
        // a replay adds no purchase; a fresh code's badge can be retired on arrival
        BadgeRedeemOutcome.Refused(
          generalGetString(if (result.newBadge) MR.strings.badges_error_badge_ended else MR.strings.badges_error_code_used)
        )
      } else {
        appPrefs.supporterBannerShown.set(true)
        BadgeRedeemOutcome.Redeemed
      }
    }
    is BadgeRedeemResult.Failed -> {
      Log.e(TAG, "apiRedeemBadgeCode: ${result.err?.string}")
      BadgeRedeemOutcome.Refused(chatModel.controller.redeemErrorText(result.err))
    }
  }

fun showCannotRedeemAlert(message: String) {
  AlertManager.shared.showAlertMsg(title = generalGetString(MR.strings.badges_error_title), text = message)
}

@Composable
fun BadgesRedeemCodeView(modalManager: ModalManager) {
  val rhId = remember { chatModel.remoteHostId() }
  val code = remember { mutableStateOf(TextFieldValue("")) }
  val canonicalCode = remember { mutableStateOf<String?>(null) }
  val submitting = remember { mutableStateOf(false) }

  // when the text is unchanged, the field's own value is kept: it carries the cursor position and the
  // keyboard's composition state, which BasicTextField loses unless they are passed back to it
  fun applyCodeInput(v: TextFieldValue) {
    val formatted = formatBadgeCodeInput(v.text)
    code.value = if (formatted != v.text) TextFieldValue(formatted, selection = TextRange(formatted.length)) else v
    canonicalCode.value = parseBadgeCode(formatted)
  }

  fun redeem() {
    val sending = canonicalCode.value ?: return
    val user = chatModel.currentUser.value ?: return
    submitting.value = true
    withBGApi {
      val outcome = redeemBadgeCode(rhId, user, sending)
      withContext(Dispatchers.Main) {
        submitting.value = false
        when (outcome) {
          is BadgeRedeemOutcome.Redeemed -> modalManager.closeModal()
          is BadgeRedeemOutcome.Refused -> showCannotRedeemAlert(outcome.message)
          is BadgeRedeemOutcome.Cancelled -> {}
        }
      }
    }
  }

  ColumnWithScrollBar(
    Modifier.padding(horizontal = 25.dp).padding(top = 8.dp, bottom = 20.dp),
    verticalArrangement = Arrangement.spacedBy(16.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
    maxIntrinsicSize = true,
  ) {
    Text(
      stringResource(MR.strings.badges_redeem_code_title),
      style = MaterialTheme.typography.h1,
      fontWeight = FontWeight.Bold,
      color = MaterialTheme.colors.primary,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    Text(
      stringResource(MR.strings.badges_redeem_code_body),
      style = MaterialTheme.typography.body1,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    CodeField(code, submitting.value, ::applyCodeInput)

    PasteButton(submitting.value, ::applyCodeInput)

    if (appPlatform.isAndroid) {
      QRCodeScanner(padding = PaddingValues(start = 16.dp, top = 12.dp, end = 16.dp)) { text ->
        val formatted = formatBadgeCodeInput(text)
        when {
          submitting.value -> false
          parseBadgeCode(formatted) == null -> {
            AlertManager.shared.showAlertMsg(
              title = generalGetString(MR.strings.invalid_qr_code),
              text = generalGetString(MR.strings.badges_code_you_scanned_is_not_badge_code)
            )
            false
          }
          else -> {
            applyCodeInput(TextFieldValue(formatted, selection = TextRange(formatted.length)))
            redeem()
            true
          }
        }
      }
    }

    Spacer(Modifier.weight(1f))

    Column(horizontalAlignment = Alignment.CenterHorizontally) {
      SubmitButton(enabled = canonicalCode.value != null && !submitting.value, onClick = ::redeem)
      TextButtonBelowOnboardingButton("", null)
    }
  }

  if (submitting.value) {
    Box(
      Modifier.fillMaxSize(),
      contentAlignment = Alignment.Center
    ) {
      Surface(Modifier.size(50.dp), color = MaterialTheme.colors.background.copy(0.9f), contentColor = LocalContentColor.current, shape = RoundedCornerShape(50)){}
      CircularProgressIndicator(
        Modifier
          .padding(horizontal = 2.dp)
          .size(30.dp),
        color = MaterialTheme.colors.secondary,
        strokeWidth = 3.dp
      )
    }
  }
}

@Composable
private fun CodeField(code: MutableState<TextFieldValue>, submitting: Boolean, applyCodeInput: (TextFieldValue) -> Unit) {
  val colors = TextFieldDefaults.textFieldColors(
    backgroundColor = MaterialTheme.appColors.sentMessage,
    textColor = MaterialTheme.colors.onBackground,
    focusedIndicatorColor = Color.Unspecified,
    unfocusedIndicatorColor = Color.Unspecified,
  )
  BasicTextField(
    value = code.value,
    onValueChange = applyCodeInput,
    enabled = !submitting,
    singleLine = true,
    textStyle = TextStyle.Default.copy(
      color = MaterialTheme.colors.onBackground,
      fontFamily = FontFamily.Monospace,
      fontSize = MaterialTheme.typography.body1.fontSize,
      textAlign = TextAlign.Center
    ),
    keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Characters, autoCorrect = false),
    cursorBrush = SolidColor(colors.cursorColor(false).value),
    modifier = Modifier
      .fillMaxWidth()
      .clip(RoundedCornerShape(10.dp))
      .background(colors.backgroundColor(!submitting).value)
      .padding(start = 12.dp, top = 14.dp, end = 12.dp, bottom = 14.dp),
    decorationBox = { innerTextField ->
      Box(contentAlignment = Alignment.Center) {
        if (code.value.text.isEmpty()) {
          Text(stringResource(MR.strings.badges_redeem_code_placeholder), color = MaterialTheme.colors.secondary, fontFamily = FontFamily.Monospace)
        }
        innerTextField()
      }
    }
  )
}

@Composable
private fun PasteButton(submitting: Boolean, applyCodeInput: (TextFieldValue) -> Unit) {
  val clipboard = LocalClipboardManager.current
  TextButton(
    onClick = { clipboard.getText()?.text?.let { applyCodeInput(TextFieldValue(it, selection = TextRange(it.length))) } },
    enabled = !submitting
  ) {
    Text(stringResource(MR.strings.paste_button), color = MaterialTheme.colors.primary, fontWeight = FontWeight.Medium)
  }
}

@Composable
private fun SubmitButton(enabled: Boolean, onClick: () -> Unit) {
  OnboardingActionButton(
    modifier = if (appPlatform.isAndroid) Modifier.padding(horizontal = DEFAULT_ONBOARDING_HORIZONTAL_PADDING).fillMaxWidth() else Modifier.widthIn(min = 300.dp),
    labelId = MR.strings.badges_redeem,
    onboarding = null,
    enabled = enabled,
    onclick = onClick
  )
}

fun openBadgeLink(rhId: Long?, codeText: String) {
  val code = parseBadgeCode(codeText)
    ?: return showCannotRedeemAlert(generalGetString(MR.strings.badges_error_invalid_code))
  ModalManager.end.showCustomModal { close ->
    BadgesRedeemLinkView(rhId, code, close)
  }
}

enum class BadgeLinkStep {
  Confirming,
  Issuing,
  Redeemed
}

// Any web page or chat message can send a badge link, and a profile holds one badge at a time,
// so this screen asks before redeeming, names the profile, and offers nothing but the redemption.
@Composable
fun BadgesRedeemLinkView(rhId: Long?, code: String, close: () -> Unit) {
  val step = remember { mutableStateOf(BadgeLinkStep.Confirming) }

  fun redeemFromLink() {
    val user = chatModel.currentUser.value ?: return close()
    step.value = BadgeLinkStep.Issuing
    withBGApi {
      val outcome = redeemBadgeCode(rhId, user, code)
      withContext(Dispatchers.Main) {
        when (outcome) {
          is BadgeRedeemOutcome.Redeemed -> step.value = BadgeLinkStep.Redeemed
          is BadgeRedeemOutcome.Refused -> {
            close()
            showCannotRedeemAlert(outcome.message)
          }
          is BadgeRedeemOutcome.Cancelled -> close()
        }
      }
    }
  }

  when (step.value) {
    BadgeLinkStep.Confirming -> ModalView(close) { Confirming(onConfirm = ::redeemFromLink, onCancel = close) }
    BadgeLinkStep.Issuing -> ModalView(close) { BeingIssued() }
    BadgeLinkStep.Redeemed -> BadgesView(ModalManager.end, close)
  }
}

@Composable
private fun Confirming(onConfirm: () -> Unit, onCancel: () -> Unit) {
  ColumnWithScrollBar(
    Modifier.padding(horizontal = 25.dp).padding(top = 8.dp, bottom = 20.dp),
    verticalArrangement = Arrangement.spacedBy(16.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
    maxIntrinsicSize = true,
  ) {
    Text(
      stringResource(MR.strings.badges_link_confirm_title),
      style = MaterialTheme.typography.h1,
      fontWeight = FontWeight.Bold,
      color = MaterialTheme.colors.primary,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    Text(
      String.format(stringResource(MR.strings.badges_link_confirm_profile), chatModel.currentUser.value?.displayName ?: ""),
      style = MaterialTheme.typography.body1,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    Spacer(Modifier.weight(1f))

    Column(horizontalAlignment = Alignment.CenterHorizontally) {
      OnboardingActionButton(
        modifier = if (appPlatform.isAndroid) Modifier.padding(horizontal = DEFAULT_ONBOARDING_HORIZONTAL_PADDING).fillMaxWidth() else Modifier.widthIn(min = 300.dp),
        labelId = MR.strings.badges_link_add_badge,
        onboarding = null,
        onclick = onConfirm
      )
      TextButtonBelowOnboardingButton(stringResource(MR.strings.cancel_verb), onCancel)
    }
  }
}

@Composable
private fun BeingIssued() {
  ColumnWithScrollBar(
    Modifier.padding(horizontal = 25.dp).padding(top = 8.dp, bottom = 20.dp),
    verticalArrangement = Arrangement.spacedBy(16.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
    maxIntrinsicSize = true,
  ) {
    Text(
      stringResource(MR.strings.badges_being_issued),
      style = MaterialTheme.typography.h1,
      fontWeight = FontWeight.Bold,
      color = MaterialTheme.colors.primary,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    Spacer(Modifier.weight(1f))

    CircularProgressIndicator(
      Modifier.size(30.dp),
      color = MaterialTheme.colors.secondary,
      strokeWidth = 3.dp
    )

    Spacer(Modifier.weight(1f))
  }
}
